File: web/src/api/tunnel.ts
import api from './index'

export interface Tunnel {
  id?: number
  tunnel_id: string
  name: string
  type: string
  transport_mode: string
  local_addr: string
  remote_addr: string
  obfuscation_mode: string
  obfuscation_config?: string
  e2e_enabled: boolean
  tls_fingerprint: boolean
  tls_profile: string
  status: string
  node_id: string
  bytes_in?: number
  bytes_out?: number
  created_at?: string
  updated_at?: string
}

export interface CreateTunnelRequest {
  name: string
  type: string
  transport_mode?: string
  local_addr?: string
  remote_addr?: string
  obfuscation_mode?: string
  obfuscation_config?: string
  e2e_enabled?: boolean
  tls_fingerprint?: boolean
  tls_profile?: string
  node_id: string
}

export interface UpdateTunnelRequest {
  name?: string
  type?: string
  transport_mode?: string
  local_addr?: string
  remote_addr?: string
  obfuscation_mode?: string
  obfuscation_config?: string
  e2e_enabled?: boolean
  tls_fingerprint?: boolean
  tls_profile?: string
  status?: string
}

export interface TunnelStats {
  tunnel_id: string
  bytes_in: number
  bytes_out: number
  status: string
}

export const tunnelApi = {
  list: async (): Promise<Tunnel[]> => {
    const res = await api.get('/tunnels')
    return res.data || []
  },

  get: async (id: number): Promise<Tunnel> => {
    const res = await api.get(`/tunnels/${id}`)
    return res.data
  },

  create: async (data: CreateTunnelRequest): Promise<Tunnel> => {
    const res = await api.post('/tunnels', data)
    return res.data
  },

  update: async (id: number, data: UpdateTunnelRequest): Promise<void> => {
    await api.put(`/tunnels/${id}`, data)
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/tunnels/${id}`)
  },

  toggle: async (id: number): Promise<{ status: string }> => {
    const res = await api.post(`/tunnels/${id}/toggle`)
    return res.data
  },

  stats: async (id: number): Promise<TunnelStats> => {
    const res = await api.get(`/tunnels/${id}/stats`)
    return res.data
  }
}

export default tunnelApi

File: admin/internal/model/tunnel.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type Tunnel struct {
	ID              uint           `gorm:"primarykey" json:"id"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`

	TunnelID        string         `gorm:"uniqueIndex;size:64;not null" json:"tunnel_id"`
	Name            string         `gorm:"size:255" json:"name"`
	Type            string         `gorm:"size:32;not null" json:"type"`
	TransportMode   string         `gorm:"size:16;not null" json:"transport_mode"`

	LocalAddr       string         `gorm:"size:255" json:"local_addr"`
	RemoteAddr      string         `gorm:"size:255" json:"remote_addr"`
	ObfuscationMode string         `gorm:"size:32;default:none" json:"obfuscation_mode"`
	ObfuscationCfg  string         `gorm:"type:text" json:"obfuscation_config"`

	E2EEnabled      bool           `gorm:"default:true" json:"e2e_enabled"`
	TLSFingerprint  bool           `gorm:"default:false" json:"tls_fingerprint"`
	TLSProfile      string         `gorm:"size:32" json:"tls_profile"`

	Status          string         `gorm:"size:16;default:active" json:"status"`
	NodeID          string         `gorm:"size:64;index" json:"node_id"`

	BytesIn         int64          `gorm:"default:0" json:"bytes_in"`
	BytesOut        int64          `gorm:"default:0" json:"bytes_out"`
}

func (Tunnel) TableName() string {
	return "tunnels"
}